plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -117,65 +136,99 @@ lensProvider
   CodeLensParams {_textDocument = TextDocumentIdentifier {_uri}}
     -- VSCode uses URIs instead of file paths
     -- haskell-lsp provides conversion functions
-    | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri = liftIO $
+    | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri _uri =
       do
-        mbMinImports <- runAction "MinimalImports" state $ useWithStale MinimalImports nfp
+        mbMinImports <- liftIO $ runAction "MinimalImports" state $ useWithStale MinimalImports nfp
         case mbMinImports of
-          -- Implement the provider logic:
-          -- for every import, if it's lacking a explicit list, generate a code lens
-          Just (MinimalImportsResult minImports, posMapping) -> do
-            commands <-
-              sequence
-                [ generateLens pId _uri edit
-                  | (imp, Just minImport) <- minImports,
-                    Just edit <- [mkExplicitEdit pred posMapping imp minImport]
-                ]
-            return $ Right $ InL $ catMaybes commands
+          Just (MinimalImportsResult minImports _ _, posMapping) -> do
+            let lens = [ generateLens _uri curRange int
+                       | (range, int) <- minImports
+                       , let Just curRange = toCurrentRange posMapping range]
+            return $ Right $ InL lens
           _ ->
             return $ Right $ InL []
     | otherwise =
       return $ Right $ InL []
+  where generateLens :: Uri  -> Range -> Int -> CodeLens
+        generateLens uri range int =
+          CodeLens { _data_ = Just $ toJSON $ ResolveOne uri int
+                   , _range = range
+                   , _command = Nothing }
+
+lensResolveProvider :: PluginMethodHandler IdeState Method_CodeLensResolve
+lensResolveProvider ideState plId cl@(CodeLens {_data_ = Just data_}) = pluginResponse $ do
+  case fromJSON data_ of
+    Success (ResolveOne uri int) -> do
+      nfp <- getNormalizedFilePath uri
+      (MinimalImportsResult _ _ resolveMinImp) <-
+        handleMaybeM "Unable to run Minimal Imports"
+        $ liftIO
+        $ runAction "MinimalImports" ideState $ use MinimalImports nfp
+      target <- handleMaybe "Unable to resolve lens" $ resolveMinImp IM.!? int
+      let updatedCodeLens = cl & L.command ?~  mkCommand plId uri target data_
```

Comment:
Interesting, so we really split things into three here:
- First we work out where the lenses are
- Then, in resolve we work out the title
- Then, in the command handler we work out the edit

And all of them actually use `MinimalImports`. I wonder if we need to shuffle the work around a bit - if we end up doing most of the work in `MinimalImports`, then maybe we're not saving much with the resolve provider. Not sure.

Reply:
There's not that much we can delay. I guess we can delay printing the expression till when we get the title, which currently isn't done, mainy because that's what the old code did, and I havent changed it. The key thing is that for most plugins the tricky part is finding out where we need to provide it, and that always needs to be done as the first step, regardless of resolve. Especially for codeLens, there is no way to avoid computing the whole documents codeLens locations every edit.